src/test/java/com/puppycrawl/tools/checkstyle/bdd/BddParser.java

```diff
@@ -0,0 +1,133 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.bdd;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+
+public final class BddParser {
+
+    /** Stop instances being created. **/
+    private BddParser() {
+    }
+
+    /**
+     * Parses the input file provided.
+     *
+     * @throws CheckstyleException if unable to read file or file not formatted properly.
+     */
+    public static ParsedInput parse(File inputFile) throws CheckstyleException {
+        final Path filePath = Paths.get(inputFile.getAbsolutePath());
```

Comment:
lets change flow instead of reading full file and iterating lines several times, we can read line by line and process it instantly

it is not the best solution but we can start from it

pseudocode:
```
if (firstLineIsNotComment) {
 throw SomeException("input file should start with comment")
}

state = FETCH_CHECK_NAME
while (not eof file) { 
   str = readLine
   if (isClosingComment()) {
      state = FETCH_VIOLATIONS
   } else
   if (state == FETCH_CHECK_NAME) {
     builder.setCheckName(extractCheckName(str));
     state = FETCH_PROPERTIES;
   } else 
   if (state == FETCH_PROPERTIES) {
      builder.addProperty(extractPropertyNameValueAndIsDefault(str));
   }
}
```

builder is object for creating ParsedInput object

Reply:
@timurt I get the point. However, I feel we should iterate it separately for violations since we'll be looking at cases like this https://github.com/checkstyle/checkstyle/pull/10090#discussion_r648796820 in future.
I understand the config should be a single iteration (both default and non-default).